iphone/Maps/Core/iCloud/DefaultLocalDirectoryMonitor.swift

```diff
@@ -0,0 +1,215 @@
+      timer.fireDate = Date(timeIntervalSinceNow: debounceTimeInterval)
+      // Stay in the `.debounce` state.
+    case .stopped:
+      // This can happen if the read source fired and enqueued a block on the
+      // main queue but, before the main queue got to service that block, someone
+      // called `stop()`.  The correct response is to just do nothing.
+      break
+    }
+  }
+
+  private func debounceTimerDidFire() {
+    LOG(.debug, "LocalMonitor: Debounce timer did fire.")
+    guard state == .started else { return }
+    guard case .debounce(let dirSource, let timer) = dispatchSourceDebounceState else { fatalError() }
+    timer.invalidate()
+    dispatchSourceDebounceState = .started(dirSource: dirSource)
+
+    do {
+      let contents = try fileManager.contentsOfDirectory(at: directory, includingPropertiesForKeys: resourceKeys, options: [.skipsHiddenFiles])
+      let contentMetadataItems = LocalContents(contents.compactMap { url in
+        do {
+          let metadataItem = try LocalMetadataItem(fileUrl: url)
+          return metadataItem
+        } catch {
+          delegate?.didReceiveLocalMonitorError(error)
+          return nil
+        }
+      })
+
+      if !didFinishGatheringIsCalled {
+        didFinishGatheringIsCalled = true
+        LOG(.debug, "LocalMonitor: didFinishGathering called.")
+        LOG(.debug, "LocalMonitor: contentMetadataItems count: \(contentMetadataItems.count)")
+        delegate?.didFinishGathering(contents: contentMetadataItems)
+      } else {
+        LOG(.debug, "LocalMonitor: didUpdate called.")
+        LOG(.debug, "LocalMonitor: contentMetadataItems count: \(contentMetadataItems.count)")
+        delegate?.didUpdate(contents: contentMetadataItems)
+      }
+    } catch {
+      fatalError("Error while reading directory: \(error)")
+    }
+  }
+
+  private func suspendDispatchSource() {
+    if !dispatchSourceIsSuspended {
+      LOG(.debug, "LocalMonitor: Suspend dispatch source.")
+      dispatchSource?.suspend()
+      dispatchSourceIsSuspended = true
+      dispatchSourceIsResumed = false
+    }
+  }
+
+  private func resumeDispatchSource() {
+    if !dispatchSourceIsResumed {
+      LOG(.debug, "LocalMonitor: Resume dispatch source.")
+      dispatchSource?.resume()
+      dispatchSourceIsResumed = true
+      dispatchSourceIsSuspended = false
+    }
+  }
+}
+
+private extension DefaultLocalDirectoryMonitor.DispatchSourceDebounceState {
+  var isRunning: Bool {
+    switch self {
+    case .stopped: return false
+    case .started: return true
+    case .debounce: return true
+    }
+  }
+}
+
+private extension FileManager {
+  func source(for directory: URL) throws -> DispatchSourceFileSystemObject {
+    if !fileExists(atPath: directory.path) {
+      do {
+        try createDirectory(at: directory, withIntermediateDirectories: true)
+      } catch {
+        throw error
+      }
+    }
+    let directoryFileDescriptor = open(directory.path, O_EVTONLY)
```

Comment:
This code monitor only directory changes, not file changes. E.g. if you edit/overwrite some file in-place, it won't fire. If you add a new file, delete, change file name, then it fires.

Our bookmarks saving code should create a new temporary file, and move it to overwrite the old file atomically, that's why it works.

However, there could be some other cases when files are just overwritten/appended, without modifying the directory.

Another case to test: what if user uploads some good/bad kml files (and other types too) using Finder/File Sharing into OM's monitored folder?

Reply:
I wrote the test case to check a file contents update isolated from the bookmark manager.
This code is triggered on _every write operation_ into the directory, not only the creation/renaming/removal.

<img width="806" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/feef7adb-0639-4fd8-940c-0175e11e365d">



> Another case to test: what if user uploads some good/bad kml files (and other types too) using Finder/File Sharing into OM's monitored folder?

We can only add files to the iCloud local container.

- Good kmls will be parsed and displayed immediately
- Bad kmls will not be displayed because they will not be parsed - I mentioned in todo that we should add an alert to handle this situation and notify a user.
- Other files (not *.kml) will not be synced (until we implement the support of the nested files/directories).

Reply:
Is any appending write to the existing file also catched? Can a test for append write be added?

Is it possible to avoid detection of appending data?

Mu Files app shows non-kml files. Is it a bug that they are synced to iCloud?

![image](https://github.com/organicmaps/organicmaps/assets/170263/db5055d0-ffd6-4f88-9d8a-91ac5880eb9d)

Reply:
> Is any appending write to the existing file also catched? Can a test for append write be added?

is this not appending writing?
<img width="279" alt="image" src="https://github.com/organicmaps/organicmaps/assets/79797627/a4992459-0ac2-4617-9760-c215b5f80af6">

> Is it possible to avoid detection of appending data?

Why should we do that?

> My Files app shows non-kml files. Is it a bug that they are synced to iCloud?

It seems that these are temporary kmls created by the core...
I'll add filtering to these files.